File: leetcode/cpp/1371.find-the-longest-substring-containing-vowels-in-even-counts.cpp
class Solution {
public:
    int findTheLongestSubstring(string s) {
        int mask = 0;
        unordered_map<int, int> prefix_index = {
            {0, -1}
        };
        string vowels = "aeiou";
        int i = 0;
        int ans = 0;
        for (const char c: s) {
            if (vowels.find(c) != std::string::npos) {
                mask ^= (1 << (c - (int)'a'));
            }
            if (prefix_index.find(mask) != prefix_index.cend()) {
                ans = max(ans, i - prefix_index[mask]);
            } else {
                prefix_index[mask] = i;
            }
            i++;
        }
        return ans;
    }
};

File: leetcode/python/1371.find-the-longest-substring-containing-vowels-in-even-counts.py
class Solution:
    def findTheLongestSubstring(self, s: str) -> int:
        mask = 0
        previous_mask = {0: -1}
        ans = 0
        for i,c in enumerate(s):
            if c in "aeiou":
                mask = mask ^ (1 << (ord(c) - ord('a')))
            if mask in previous_mask:
                ans = max(ans, i - previous_mask[mask])
            else:
                previous_mask[mask] = i
        return ans

